### ISSUE TITLE
Add ImpersonatedCredentials 

### ISSUE THREAD

**Issue by salrashid123** [googleapis/google-api-go-client#378]
Add capability to impersonate a service account.

This feature is idential to the capability existing in 
- [google-auth-python](https://google-auth.readthedocs.io/en/latest/reference/google.auth.impersonated_credentials.html)
- [google-auth-java: ImpersonatedCredentials.java](https://github.com/googleapis/google-auth-library-java/blob/master/oauth2_http/java/com/google/auth/oauth2/ImpersonatedCredentials.java)

the original PR against oauth2 was abandoned on suggestion it gets added to this repo
ref: https://github.com/golang/oauth2/issues/336

as  a folwlup, i created https://github.com/googleapis/google-api-go-client/pull/337  againts this repo but on request from the maintainers here back then,  they would implement it separately somehow.   Its not implemented yet so are we ok to reopen the the PR 337?  

If so, the sample usage looks like
https://gist.github.com/salrashid123/7994be642c466844f85e2f2c68a5dec9

**Comment by jadekler** [comment]
Hey @salrashid123! We're working on this, but it's going to be a while. It's moderately tricky to make this work properly, and as users expect in all the corner cases, and how it fits in with all the other credentials.

**Comment by salrashid123** [comment]
until its figured out and implemented:
 an  impersonatedcredentials and a way to get idTokens.  it only works if credential object is supplied

- https://github.com/salrashid123/oauth2

**Comment by Bowbaq** [comment]
@jadekler Is there any update on this. I'm running into https://issuetracker.google.com/issues/142086394 and it seems that the only fix available to me would be to impersonate a user

**Comment by salrashid123** [comment]
@Bowbaq i dont' think you're after the same impersonation as requested here.  If you're aftr gsuites "delegated access" (eg, for drive, spreadsheets, etc), see

https://gist.github.com/salrashid123/7994be642c466844f85e2f2c68a5dec9#gistcomment-2968833

**Comment by Bowbaq** [comment]
Actually, the reporter of that issue uses G Suite, but I don't, though I also run into the same problem with emails not being sent documents are shared by a service account.

**Comment by gaahrdner** [comment]
I too, desperately need this.

**Comment by shellcromancer** [comment]
+1 to adding this feature. Would be super helpful with managing IAM as Code and restricting access in GCP

**Comment by salrashid123** [comment]
from a comment a bit above, i've put together `ImpersonatedCredential` TokenSource here
- https://github.com/salrashid123/oauth2#usage-impersonatedcredentials

and to explicitly state, what's there isn't officially google supported.

**Comment by gaahrdner** [comment]
Awesome work @salrashid123, how do we go about getting this in mainline?

**Comment by salrashid123** [comment]
thats actually a question for the maintainers.

@JustinBeckwith @jadekler @tbpg 
 - any updates on the redesign thats mentioned back in june?

**Comment by jadekler** [comment]
Unfortunately not yet. We're working on putting together a team to work on auth things like this.

**Comment by logsnitch** [comment]
I've spent literally days trying to get this working and have gotten nowhere. It is wild that there is no built in support for impersonating accounts

**Comment by derekperkins** [comment]
If this issue is going to remain open and unsolved, could it be pushed into the documentation somewhere? Like most people here, it took us a few days to figure it out.

**Comment by salrashid123** [comment]
just to be clear, the FR here is about  a user or serviceaccount impersonating _another_ serviceAccount  (i.,e using [this feature](https://cloud.google.com/iam/docs/creating-short-lived-service-account-credentials) for GCP IAM)

if on the other hand you need a serviceAccount to impersonate a Gsuies/Cloud Identity _user_, thats a different api and here is a snippet for that

https://gist.github.com/salrashid123/43dfed22880ac91a079818da7ba549c0
(eg, the bit above uses a serivce account to impersonate a Gsuites user and then as that user access drive, gcs, and list users if that user has/had permissions to that)

**Comment by derekperkins** [comment]
@salrashid123 Thanks for that, we are indeed looking for user impersonation. We attempted using your library (https://github.com/salrashid123/oauth2#usage-impersonatedcredentials) and kept getting insufficient scope errors, even though we granted `Owner` access to the SA in question.

We also attempted using this snippet from @broady https://github.com/googleapis/google-api-go-client/issues/379#issuecomment-512343632, which at first glance appears to be the same as the gist you just helpfully supplied. We'll give it a go again after looking at your example in more detail.

What we specifically want to work is to use metadata server provided credentials (Workload Identity in GKE) to impersonate a user. That is even more difficult to do as you mentioned here: https://github.com/googleapis/google-api-go-client/issues/379#issuecomment-513860042

Issue #379 about user impersonation was closed in favor of this issue. @codyoss @jadekler Should we create a new issue for a service account impersonating a G Suite user via Domain Wide Delegation, or does the scope of this issue include that as well?

Beyond just adding support to this SDK, it would also be very helpful if the server could provide more helpful error responses, as the other traditional IAM calls do. All we see now is `Insufficient scopes provided`, seemingly no matter what scopes we request or what privileges we provide the SA.

**Comment by salrashid123** [comment]
so, go does have SDK for the `SA->user` impersonaton  per #379   ...Its just that in the current form of that implementation the Service account necessarily needs to be in is a json file.  The json cert file is used there to sign a jwt with a specific claim that signifies the user to impersonate.  Then the JWT is excahnged with google to get the users `access_token`

if anythihing the #379 should be reopened to support other sources of service account based tokens.

Workload Identity and Computes metadta simply gives you a token and that alone isn't what the dwd flow requires (as mentioned you to send a signedJWT as the SA to google and then get the users token back).  

using https://github.com/salrashid123/oauth2#usage-impersonatedcredentials wont' work there no matter the scopes since it's `SA->SA` impersonation only when running in compute (this FR is for the link in that article to be part of the official library!)

----

If its helpful, here is the j[ava implementation](https://medium.com/@salmaan.rashid/easy-gsuites-domain-wide-delegation-dwd-in-java-5abc7d08f1e1) of what you're after:

specifically here i'm talkign a credential from GCE and using it to sign a JWT where the `sub` field is the user to impersonate via dwd
https://gist.github.com/salrashid123/d60987fe367b137833281732922c31f9#file-domaindelegatedcredentials-java-L122

**Comment by salrashid123** [comment]
oh, wait, i forgot i implemented the IAM based signed creds into impersonatedcredentials:
   https://github.com/salrashid123/oauth2#impersonated-credentials-with-domain-wide-delegation

In your case, you need to allow ServiceAccount that WorkloadIdentity runs as the "TokenCreator" role on itself (inorder to use the IAM SignJWT API).  Once you do that, the credential you need to instantiate isn't just application default credentials but an impersonated credential tokensource derived from ADC.  While instantiating the impersonatedcreds, speciy the `subject`  field and then use those creds in an api.

the dwd capability is pretty intense (even more than the baseline `sa->sa` impersoantion per this issue)... so needless to say, pls be careful !

as before, i'd suggest forkign the implementation of the impersonated creds off of that link (the stuff i have is ofcourse unsupported and may break)

**Comment by fredzqm** [comment]
@jadekler 
Hi Jean,
It looks like this feature has slipped away from the priority.
Really want to help raise attention and priority for this because impersonating really opens up doors to s/acl-based/capability-based/ security.
I am working in a G-on-G project that can benefit greatly from this feature.

Thank you for the community contribution you made here.
I am hoping that you can help escalate the priority of this FR.

**Comment by jadekler** [comment]
cc @codyoss @bshaffer

**Comment by codyoss** [comment]
Once we land some of the ID Token work that is underway, I think we will have an easier path to impersonation. I will make sure this gets re-brought up in internal discussions. I know it is still on peoples minds.

**Comment by yhuang** [comment]
What can current GCP customers do to help escalate the priority of this PR?  @salrashid123, I will reach out via my TAM, so we can meet soon to discuss a way forward to close this gap.

**Comment by codyoss** [comment]
Thanks for your interest in the feature. I think reaching out through your TAM is a good way to escalate.

**Comment by khipkin** [comment]
Is there an internal issue tracking this FR that we can +1? i.e. on https://issuetracker.google.com.

**Comment by yhuang** [comment]
@khipkin: I have opened a private partner buganizer issue and started the escalation with my TAM already.

This particular hurdle with an SDK that is in Google's own Go is hitting a nerve, because we as customers, again, are being forced to work harder to advocate for things that should already be in place.  Last year, we had to advocate for service account impersonation to be supported by gsutil, because there is no overall mandate at Google to make sure support for that two-year-old feature is uniform across CLI tools and all language SDKs. 

Google should have more alignment from all the CLI and SDK development teams, so once a security foundation feature like this becomes publicly available in the API, customers can take advantage of the capabilities in the language of their choice.  The current approach is far from optimal.

**Comment by carnott-snap** [comment]
I have implemented in [CL 214477](https://go-review.googlesource.com/c/oauth2/+/214477), but the review has stalled. If it should live here, I think all the symbols are exported, so I can port the logic here. It would be nice to use the actual IAM lib, but that is not part of the API, so I can fixup the implementation later. Otherwise, can I get some help driving CL 214477?

**Comment by LewisLebentz** [comment]
I'm trying to create a tool that will allow me to add delegates to Gmail, but seem to be having a similar issue.

I've granted a service account DwD, but can't seem to be able to get Go to use the creds, or impersonate a user so I can modify the Gmail settings.

@salrashid123 I saw your sample for Drive, but it looks like it isn't possible to modify Gmail settings for G Suite users using Go at the moment? If I'm missing something, then great! But I can't find much else about it on the Internet other than this ticket. If necessary I will just use Python.

**Comment by carnott-snap** [comment]
can you try git.sr.ht/~urandom/dwd? I published my CL as a package, and that should let you do what you need. If it works we can lean on getting it merged, or maybe put it in this repo, if the core team does not want it in oauth2.

**Comment by LewisLebentz** [comment]
@carnott-snap not great at Go yet, trying to use it with this API:

https://developers.google.com/gmail/api/v1/reference/users/settings/delegates

How would I go about using your library to auth for the delegates API?

**Comment by carnott-snap** [comment]
Mail delegation should be orthogonal to the actual token delegation. E.g. I used that lib to interact with the directory api, where my service account needed to impersonate my google account: https://developers.google.com/admin-sdk/directory/v1/guides/delegation

**Comment by codyoss** [comment]
An option has been added to support user/service-account impersonation of another service account. I created a new separate issue, linked above, to track supporting impersonating a gSuite admin user.

**Comment by Sytten** [comment]
It is not really how this should be used and I can't find any example anywhere. I am trying to make it work on my dev machine where I am using user credentials, but I always get a 403:
```
unknown error while making an http call: Post \"https://identitytoolkit.googleapis.com/v1/projects/REDACTED/accounts:lookup\": impersonate: status code 403: {\n  \"error\": {\n    \"code\": 403,\n    \"message\": \"The caller does not have permission\",\n    \"status\": \"PERMISSION_DENIED\"\n  }\n}\n
```
I put the email in the target field, is that the right thing? @codyoss

**Comment by codyoss** [comment]
Hey @Sytten Please open a new issue if you have a question related to the PR. 

Looking at the error it sounds like a scoping issue. Try passing option.WithScopes("https://www.googleapis.com/auth/cloud-platform") and see if that fixes your issue. If you are still facing issues after this please open a new issue with a reproducer.

**Comment by joshk0** [comment]
Feel free to follow #731 to get this behavior improved, @Sytten . That's exactly what I ran into as well.

**Comment by Sytten** [comment]
Thank you all for the answers, it really was not obvious. I will have to read a bit more about scopes too, I thought they were somewhat deprecated in favor of roles.

**Comment by Sytten** [comment]
Ok I found my issue on local desktop. I forgot that the gcloud config doesn't change the application default credentials and it was thus using another account causing the 403. It is really not obvious since it doesn't really print the account that was used initially in the delegation.

**Comment by codyoss** [comment]
Please try out our impersonate preview and feel free to provide any feedback: https://github.com/googleapis/google-api-go-client/releases/tag/v0.44.0-impersonate-preview

**Comment by Sytten** [comment]
How is it different than what we do currently?

**Comment by codyoss** [comment]
> How is it different than what we do currently?

It is a little different in that the new package gives you a TokenSource to work with that you will then pass into the client with a client option, [see example](https://pkg.go.dev/google.golang.org/api@v0.44.0-impersonate-preview/impersonate#example-CredentialsTokenSource-ServiceAccount). It provides more flexibility particularly when constraining to tighter scopes. The package also support creating impersonated idtokens and allows for impersonating admin user with domain wide delegation.

**Comment by Sytten** [comment]
Thanks @salrashid123 that is exactly what I wanted!

**Comment by salrashid123** [comment]
just one note on the preview implementation...   I realize that in normal cases it doens't make sense to set a `duration` field while getting an impersonated token because the auth client would autorefresh it anyway (i.,e if in the example above if you set it to  expire in 120s, the client lib woud just refresh it anyway automatically so might as well not set it and use the default `3600s`)

however, i could  ask for a raw  impersonated `access_token` and then [downscope](https://cloud.google.com/iam/docs/downscoping-short-lived-credentials#exchange-credential) it to attenuate the resources it has access to (eg, to just a set of gcs buckets)...in that case, i'd also like to restricts how long its valid for with the full understanding its non-refreshable.   Basically i'd like a token that represents another service account but is restricted to certain resource _and_ is valid for only `60s`.

 to do that i'd need the original token to be limited to `60s` (in this case the original token is the one derived from impersonation)

**Comment by codyoss** [comment]
@salrashid123 

> i'd also like to restricts how long its valid for with the full understanding its non-refreshable

If the duration of the token is configured impersonated token will not be auto-refreshed. So I think this covers what you are asking about? 
https://github.com/googleapis/google-api-go-client/blob/impersonate-preview/impersonate/impersonate.go#L80

**Comment by salrashid123** [comment]
its a bit different than that:

1. user U impersonates svc_account S1.
2. S1 has access to GCS buckets A,B,C
3. user U gets the raw access token T1  for S1 
4. user U takes T1 and then downscopes its permission to have access to only bucket A and gets token T2
5. user gives T2 to some other system...at that point, T2 still represents S1 but has access to only bucket A

the think i''m pointing out is that there T1 is always set to `3600s` and consequently T2 is always set to `3600s`.

if in step 1 the user can stipulatge that T1 is valid for only `60s`, then at the end T2 is also `60s`
